sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala

```diff
@@ -403,20 +403,30 @@ trait Row extends Serializable {
       if (!isNullAt(i)) {
         val o1 = get(i)
         val o2 = other.get(i)
-        if (o1.isInstanceOf[Array[Byte]]) {
-          // handle equality of Array[Byte]
-          val b1 = o1.asInstanceOf[Array[Byte]]
-          if (!o2.isInstanceOf[Array[Byte]] ||
-            !java.util.Arrays.equals(b1, o2.asInstanceOf[Array[Byte]])) {
+        o1 match {
+          case b1: Array[Byte] =>
+            if (!o2.isInstanceOf[Array[Byte]] ||
+              !java.util.Arrays.equals(b1, o2.asInstanceOf[Array[Byte]])) {
+              return false
+            }
+          case f1: Float =>
+            if (!o2.isInstanceOf[Float] ||
+              (java.lang.Float.isNaN(f1) && !java.lang.Float.isNaN(o2.asInstanceOf[Float]))) {
```

Comment:
actually, I don't think that we can use nanSafeCompare without breaking existing test code / user code: the old code would allow integers and floats to be compared because Java would handle implicit type conversions.  Therefore, for compatibility I think we need to do the same here.

I think it will be clearer to rework this as something like "if f1 is a float and it's NaN, then the other value had better be a NaN float, otherwise fall back to the regular `==` branch).